server.go

```diff
@@ -148,7 +148,7 @@ func (s *SSServer) loadConfig(filename string) error {
 		}
 	}
 	for portNum, cipherList := range portCiphers {
-		s.ports[portNum].cipherList = cipherList
```

Comment:
Isn't this assignment already atomic? Why do we need a mutex?

Reply:
Assignments in Go are not guaranteed to be atomic.

In practice, I don't think this will be a problem, but the safety of this code currently relies on guarantees that the language specification does not make.

Reply:
There's something off with the abstraction layering here (my fault, not yours). We are effective using a CipherList for the only purpose of creating a list.List that gets assigned to the exiting CipherList. 

Let's get rid of the intermediary CipherList and work with a simple list.List here instead. This is my suggestion:
- Change like 115 to:
```golang
portCiphers := make(map[int]list.List)
```
- Adjust the other lines as needed. Line 121 will become `cipherList = list.New()`. Remove the `PushBack` method from `cipherList`.
- Add a `SafeUpdate(list.List)` to cipherList. (assigns the list.List under the Mutex)
- Change line 151 to use `SafeUpdate`

Reply:
Done. I removed the `Safe` prefixes since all methods are now thread-safe.